### ISSUE TITLE
Potential vanishing gradient problem

### ISSUE THREAD

**Issue by rlindsberg** [YinengWang/Supervised-Image-Classification-with-Noisy-Labels-Using-Deep-Learning#11]
The gradients seem to be small enough so that the weights are not getting updated.

At step 5000, the gradients reach the end of the plots. At the same time, the accuracy stays the same. To check yourself: https://wandb.ai/yineng_w/replicate-cifar10/runs/v5uramk5?workspace=

![Skærmbillede 2021-05-14 kl  22 22 42](https://user-images.githubusercontent.com/2918707/118325418-ba4bef00-b503-11eb-9695-5fe969626f58.png)

To reproduce: run commit 0e69f39000cbcb331c5196c7a5a4bc15ad880925 for 20 epochs. The problem becomes present at epoch 13.

**Comment by YinengWang** [comment]
This running uses ELR loss. The implementation needs to be checked.

**Comment by rlindsberg** [comment]
We re-run using CE and will check again

**Comment by Alfred-N** [comment]
Ive done tests with lambda =0 (regularization parameter) and it works exactly like CE, so we know the problem is limited to the ELR-term. You can see the comparison of lambda = 0, lambda=3 here: https://wandb.ai/dd2424-group9/resnet-elr-cifar10?workspace=user-alfredn

**Comment by Alfred-N** [comment]
I've also discovered that they seem to be using CosineAnnealingWarmRestarts for ELR in the /ELR/config_cifar10.json:

`    "lr_scheduler": {
        "type": "CosineAnnealingWarmRestarts",
        "args": {
            "T_0": 10,
            "eta_min": 0.001
        }
    },
`
as well as beta=0.7 instead of 0.3 which I used previously.

Im going to try this and see what happens

**Comment by YinengWang** [comment]
> I've also discovered that they seem to be using CosineAnnealingWarmRestarts for ELR

Sure, but they also use multi-step annealing. We need to test with both.

**Comment by Alfred-N** [comment]
![image](https://user-images.githubusercontent.com/47230525/118413997-70e5d600-b6a2-11eb-97e0-091a3a4dd3a6.png)
Ok now, I don't know what's going on, it seems way too good for 40% noise

**Comment by Alfred-N** [comment]
I found the bug. The authors find clamped their activasions in the range [10^-4,1-10^-4] to avoids zero grads.
I didn't do that, it's added to our loss as well now.
Before:
![image](https://user-images.githubusercontent.com/47230525/118415686-25d0c080-b6ac-11eb-97ab-7d4dd1f417fc.png)
After:
![image](https://user-images.githubusercontent.com/47230525/118415697-34b77300-b6ac-11eb-9f46-b8b1322929a5.png)

The plots also look better now (at least test accuracy):
![image](https://user-images.githubusercontent.com/47230525/118415745-79dba500-b6ac-11eb-950b-1ad65055dd16.png)
![image](https://user-images.githubusercontent.com/47230525/118415752-83650d00-b6ac-11eb-84c1-596de2d413fb.png)

**Comment by Alfred-N** [comment]
Now I got some new wierd error unfortunately, will continue tomorrow
![image](https://user-images.githubusercontent.com/47230525/118416580-bd381280-b6b0-11eb-91b5-c94658b9a6fb.png)

**Comment by rlindsberg** [comment]
Amazing!!

**Comment by rlindsberg** [comment]
closed by #17
